torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -190,118 +190,152 @@ Node* getOrCreateTensorExprSubgraph(Node* n) {
  public:
   TensorExprFuser(std::shared_ptr<Graph> graph) : graph_(std::move(graph)) {}
 
   void run() {
     aliasDb_ = torch::make_unique<AliasDb>(graph_);
-    createFusionGroups();
+    createFusionGroups(graph_->block());
   }
 
  private:
-  // Merge fusible nodes into subgraphs in tensorexpr::Group nodes.
-  void createFusionGroups() {
-    auto block = graph_->block();
-
-    std::vector<std::pair<graph_node_list_iterator, graph_node_list_iterator>>
-        worklist;
-    std::unordered_set<torch::jit::Block*> visited_blocks;
-
-    bool any_changed = true;
-    while (any_changed) {
-      any_changed = false;
-      worklist.push_back({block->nodes().rbegin(), block->nodes().rend()});
-
-      while (worklist.size()) {
-        auto& it = worklist.back().first;
-        auto end = worklist.back().second;
-
-        if (it->blocks().size()) {
-          Node* n = *it;
-          ++it;
-
-          if (it == end) {
-            worklist.pop_back();
-          }
-
-          for (auto b : n->blocks()) {
-            if (!visited_blocks.count(b)) {
-              worklist.push_back({b->nodes().rbegin(), b->nodes().rend()});
-              visited_blocks.insert(b);
-            }
-          }
-        } else {
-          bool changed;
-          std::tie(it, changed) = scanNode(*it);
-          any_changed |= changed;
-          if (it == end) {
-            worklist.pop_back();
-          }
-        }
+  // Add unvisited input nodes to the queue for further merging into the fusion
+  // group.
+  void updateQueue(
+      Node* fusion_group,
+      std::set<Node*, nodesComparator>& queue,
+      const std::unordered_set<Node*>& visited) {
+    for (auto input : fusion_group->inputs()) {
+      if (!visited.count(input->node())) {
+        queue.insert(input->node());
       }
     }
   }
 
-  std::pair<graph_node_list::iterator, bool> scanNode(Node* consumer) {
-    auto inputs =
-        sortReverseTopological(consumer->inputs(), consumer->owningBlock());
-
-    // Grab the iterator below consumer.  We'll use that to determine
-    // where to resume iteration, even if consumer gets relocated within
-    // the block.
-    auto iter = --consumer->reverseIterator();
-    for (auto input : inputs) {
-      if (auto group = tryMerge(consumer, input->node())) {
-        // Resume iteration from where consumer is/used to be.
-        return {++iter, true};
-      }
+  // Create a fusion group starting from the node N.
+  // We then try to pull inputs into the fusion group and repeat that process
+  // until there is nothing we can pull in.
+  Node* createFusionGroup(Node* n) {
+    // Queue of the nodes we should consider for merging into the fusion groups
+    // (those nodes are usually inputs of the fusion group).
+    // We use an ordered set here to visit them in the right order: the fusion
+    // group is closer to the end of the block and we are trying to pull later
+    // nodes first.
+    std::set<Node*, nodesComparator> queue;
```

Comment:
I'm trying to think of edge cases where the queue / visited set might not behave how we want them to & get in the way of fusion... One thing is the queue might get stale/incorrect if a node gets moved around as part of  `moveBeforeTopologicallyValid`. However I think a node will only get moved if we haven't been able to merge it, so I'm not sure how much this matters. 

I guess something to think about but I can't actually think of an edge case so we should go ahead with this.

Reply:
That's another very good point and I'm very happy that you also answered your concern :) I didn't think of that but after your description it indeed looks like it would still work fine. I will add a comment highlighting that.